lib/internal/streams/compose.js

```diff
@@ -0,0 +1,216 @@
+'use strict';
+
+const pipeline = require('internal/streams/pipeline');
+const Duplex = require('internal/streams/duplex');
+const Readable = require('internal/streams/readable');
+const PassThrough = require('internal/streams/passthrough');
+const eos = require('internal/streams/end-of-stream');
+const { destroyer } = require('internal/streams/destroy');
+const assert = require('internal/assert');
+const from = require('internal/streams/from');
+const {
+  isReadable,
+  isWritable,
+  isStream,
+  isIterable
+} = require('internal/streams/utils');
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_MISSING_ARGS,
+  },
+} = require('internal/errors');
+
+class ComposeDuplex extends Duplex {
+  constructor(options) {
+    super(options);
+
+    // https://github.com/nodejs/node/pull/34385
+
+    if (options?.readable === false) {
+      this._readableState.readable = false;
+      this._readableState.ended = true;
+      this._readableState.endEmitted = true;
+    }
+
+    if (options?.writable === false) {
+      this._writableState.writable = false;
+      this._writableState.ending = true;
+      this._writableState.ended = true;
+      this._writableState.finished = true;
+    }
+  }
+}
+
+module.exports = function compose(...streams) {
+  let ondrain;
+  let onfinish;
+  let onreadable;
+  let onclose;
+  let d;
+  let head;
+  let tail;
+
+  function onfinished (err) {
+    if (onclose) {
+      onclose(err);
+    } else if (err) {
+      d.destroy(err);
+    }
+    onclose = null;
+  }
+
+  if (streams.length === 0) {
+    throw new ERR_MISSING_ARGS('streams');
+  } else if (streams.length === 1) {
+    // TODO (ronag): Move this logic into from?
+    if (typeof streams[0] === 'function') {
+      if (streams[0].length === 0) {
+        return from(ComposeDuplex, streams[0](), { writable: false });
+      }
+
```

Comment:
I see one problem here. We currently assume that if the generator has a signature with more than 1 parameter that then it's a transformer or async function, e.g.

```js
async function*(source){}
async function(source){}
```

However, in https://github.com/nodejs/node/pull/39067 I'm proposing that we should also have a signal parameter to generators.

```js
async function*(source, { signal }){}
async function(source, { signal }){}
```

Which then could lead to read only generators being detected as transformers:


```js
async function*({ signal }){}
async function({ signal }){}
```

I'm a little unsure how to get around this? Should we allow passing options with each stream argument?

```js
compose(
  { readable: true, writable: false, stream: async function *({ signal }),
  ...streams
)
```

Reply:
@mcollina @benjamingr

Reply:
This is only a problem for the first argument and read-only.

Reply:
Another option is:

read-only: signature is always length === 1
read-write: signature is always length === 2

Reply:
@benjamingr: Is the following a possibility?

```js
async function*({ source, signal }){}
async function({ source, signal }){}
```

Reply:
KISS. We don't allow read-only async generators? If you want to use it you have to do:

```js
compose(
 Readable.from(async function *({ signal })),
 async function *(source, { signal }),
 async function (source, { signal }),
)
```

Reply:
`compose` might be doing too much now, leading to ambiguities and inability to extend APIs (like adding a signal, or other parameters in the future). If streams can indicate whether they're readable/writable and async generators cannot, then `compose` should only accept streams as input. I.e.:

```js
compose(
  Readable.from(async function *({ signal })),
  Transform.from(async function *(source, { signal })),
  Writable.from(async function (source, { signal })),
)
```

Reply:
We could do it like that but it would be significantly slower.

Reply:
I think I've managed to resolve the ambiguity. We only allow passing Iterable factories which take a source. If you want to convert a read-only iterable generator to a stream you must instantiate the iterable. Updated example and docs.

Reply:
So the implementation righjt now is:

```js
async function *(...)() // readable
async function *(...) // readable & writable
async function(...) // writable
```

No ambiguity.